### DISCUSSION TITLE
/opt/oe_lvi/cc is not found while building logging app

### DISCUSSION THREAD

**Discussion by divneil-intel** [microsoft/CCF#4085] [Q&A] [answered]
I am trying a basic app `logging` on v2.0.4 and running into build issues. So, need some help to see if I am missing something very basic.

Docker File:
```
FROM mcr.microsoft.com/ccf/app/dev:2.0.4-sgx
COPY CCF/samples/apps/logging /home/logging/
RUN cd /home/logging/ && cmake .
```
Executing: `docker build -t ccf-logging .`

```
#8 1.178 -- Looking for dl library - found
#8 1.190 -- Performing Test OE_SPECTRE_MITIGATION_C_FLAGS_SUPPORTED
#8 1.215 -- Performing Test OE_SPECTRE_MITIGATION_C_FLAGS_SUPPORTED - Failed
#8 1.217 -- Performing Test OE_SPECTRE_MITIGATION_CXX_FLAGS_SUPPORTED
#8 1.242 -- Performing Test OE_SPECTRE_MITIGATION_CXX_FLAGS_SUPPORTED - Failed
#8 1.243 CMake Error at /opt/openenclave/lib/openenclave/cmake/configure_lvi_mitigation_build.cmake:73 (message):
#8 1.243   -- /opt/oe_lvi/cc is not found.
#8 1.243 Call Stack (most recent call first):
#8 1.243   /opt/openenclave/lib/openenclave/cmake/openenclave-lvi-mitigation-config.cmake:43 (configure_lvi_mitigation_build)
#8 1.243   /opt/ccf/cmake/ccf_app.cmake:66 (find_package)
#8 1.243   /opt/ccf/cmake/ccf-config.cmake:31 (include)
#8 1.243   CMakeLists.txt:16 (find_package)
```

Listing from `docker run -ti ccf-logging /bin/bash` run without `RUN cd /home/logging/ && cmake .`
```
root@ddebf8b36d8a:/# ls -l /opt/oe_lvi/
total 4712
-rwxrwxr-x 1 root root 2043416 Nov  2  2020 as
-rwxr-xr-x 1 root root     234 Jul  1 12:37 clang++-10
lrwxrwxrwx 1 root root      19 Jul  1 12:37 clang++-10_symlink -> /usr/bin/clang++-10
-rwxr-xr-x 1 root root     232 Jul  1 12:37 clang-10
lrwxrwxrwx 1 root root      17 Jul  1 12:37 clang-10_symlink -> /usr/bin/clang-10
-r-xr-xr-x 1 root root    1093 Jun 10 22:17 invoke_compiler
-rwxrwxr-x 1 root root 2768568 Nov  2  2020 ld
root@ddebf8b36d8a:/#
```

**Comment by eddyashton** [comment, marked as answer]
@divneil-intel We expect CCF apps to be built with `clang`, rather than `gcc`. Prefix your `cmake` invocation with `CC="/opt/oe_lvi/clang-10" CXX="/opt/oe_lvi/clang++-10"` (as in [this sample app](https://github.com/microsoft/ccf-app-template#build)), and the build should proceed correctly. Note this will not work in an existing build directory where any cached CMake state has selected `gcc` as the preferred compiler.

Sorry for the obscure error here. The LVI mitigation toolchain needs to shim CMake's compiler variables, which uses CMake's automatically detected preferred compiler, which is looked up extremely early in CMake configuration. I'll see if we can expose a way to set the default for apps, as we currently do for the framework code.

**Reply by divneil-intel** [reply]
Thanks @eddyashton for your prompt reply, the process goes beyond generating the makefiles now. However, the build is still failing.
I am on CCF commit: dfe98cbf164294d961277a6384796cd29b4b96ae and modified the dockerfile to:

Build environment is: Windows10 WSL2 Ubuntu + Docker Desktop

```
# Use the base image from Microsoft
FROM mcr.microsoft.com/ccf/app/dev:2.0.4-sgx

ENV APPDIR=/home/logging
ENV BUILDIR=$APPDIR/build

RUN rm -rf $APPDIR
COPY CCF/samples/apps/logging $APPDIR
RUN mkdir $BUILDIR
RUN cd $BUILDIR && CC="/opt/oe_lvi/clang-10" CXX="/opt/oe_lvi/clang++-10" cmake ..
RUN cd $BUILDIR && make
```

The following are the first few lines:
```
 > [7/7] RUN cd /home/logging/build && make:
#10 0.506 Scanning dependencies of target logging.enclave
#10 0.519 [ 11%] Building CXX object CMakeFiles/logging.enclave.dir/logging.cpp.o
#10 3.409 /home/logging/logging.cpp:219:37: error: only virtual member functions can be marked 'override'
#10 3.409       const ccf::AuthnPolicies& ap) override
#10 3.409                                     ^~~~~~~~
#10 3.409 /home/logging/logging.cpp:225:11: error: use of undeclared identifier 'CCF_APP_TRACE'
#10 3.439           CCF_APP_TRACE("BEGIN {} {}", verb.c_str(), method);
#10 3.439           ^
#10 3.439 /home/logging/logging.cpp:227:11: error: use of undeclared identifier 'CCF_APP_TRACE'
#10 3.462           CCF_APP_TRACE("END   {} {}", verb.c_str(), method);
```

With this, I am trying to create an upgrade reference, so, that existing app can be upgraded to newer version.

**Reply by eddyashton** [reply]
@divneil-intel It looks like you are trying to build the sample CCF app from `main` or a `3.0.0-dev` release against CCF version `2.0.4`. This is not expected to work, as the APIs on `main` include breaking changes. If you check out the `release/2.x` branch, the sample logging app [there](https://github.com/microsoft/CCF/tree/release/2.x/samples/apps/logging) will compile against CCF `2.0.4`.

**Reply by divneil-intel** [reply]
Yeah! my bad, should have been careful. Thanks a lot for your support.

**Reply by eddyashton** [reply]
No worries, happy to help.
